packages/data/src/index.js

```diff
@@ -284,6 +284,12 @@ export const withSelect = ( mapStateToProps ) => createHigherOrderComponent( ( W
 			 * @type {boolean}
 			 */
 			this.shouldComponentUpdate = false;
+			/**
+			 * Boolean tracking when can we run selection
+			 * The selection can be run only when the component is mounted.
+			 */
+			this.shouldRunSelection = false;
```

Comment:
I know I removed this previously @aduth but it looks like it's necessary, hard to reproduce in unit tests but without it, you notice some warnings about `setState` being called on components not mounted.

Reply:
We should just move the `subscribe` into `componentDidMount`.

Reply:
That's not possible because the subscribe of parent components should happen before children.

Reply:
>That's not possible because the subscribe of parent components should happen before children.

Why? Do we test or document this anywhere?

Reply:
Yes, I have a comment on the constructor. The comment could be made clearer.

For example:

if you have a BlockList component showing a Block component, you want the subscribe to happen first in `BlockList` to unmount the `Block` if it gets deleted before the `mapStateToProps` runs in the child

Reply:
>if you have a BlockList component showing a Block component, you want the subscribe to happen first in `BlockList` to unmount the `Block` if it gets deleted before the `mapStateToProps` runs in the child

Seems reasonable. I'll see if I can put together a unit test for this as part of #7431.